src/iosrv/sched.c

```diff
@@ -26,73 +26,953 @@
+out:
+	return spi->spi_space_pressure;
+}
+
+static int
+process_req(struct dss_xstream *dx, struct sched_request *req)
+{
+	struct sched_info	*info = &dx->dx_sched_info;
+	struct sched_pool_info	*spi = req->sr_pool_info;
+	struct sched_req_info	*sri;
+	unsigned int		 req_type = req->sr_attr.sra_type;
+
+	D_ASSERT(spi != NULL);
+	D_ASSERT(req_type < SCHED_REQ_MAX);
+
+	sri = &spi->spi_req_array[req_type];
+
+	/* Kickoff all requests on shutdown */
+	if (info->si_stop)
+		goto kickoff;
+
+	if (sri->sri_req_kicked < sri->sri_req_limit)
+		goto kickoff;
+
+	/* Request expired */
+	D_ASSERT(info->si_cur_ts >= req->sr_enqueue_ts);
+	if ((info->si_cur_ts - req->sr_enqueue_ts) > max_delay_msecs[req_type])
+		goto kickoff;
+
+	/* Remaining requests are not expired */
+	return 1;
+kickoff:
+	sri->sri_req_kicked++;
+	req_kickoff(dx, req);
+	return 0;
+}
+
+static inline void
+process_req_list(struct dss_xstream *dx, d_list_t *list)
+{
+	struct sched_request	*req, *tmp;
+	int			 rc;
+
+	d_list_for_each_entry_safe(req, tmp, list, sr_link) {
+		rc = process_req(dx, req);
+		if (rc)
+			break;
+	}
+}
+
+static inline void
+reset_req_limit(struct dss_xstream *dx, struct sched_pool_info *spi,
+		unsigned int req_type, unsigned int limit)
+{
+	unsigned int	tot = pool2req_cnt(spi, req_type);
+
+	D_ASSERT(limit <= tot);
+	if (tot - limit > max_qds[req_type]) {
+		D_CRIT("XS(%d) Too large QD: %u/%u/%u for req:%d\n",
+		       dx->dx_xs_id, tot, max_qds[req_type], limit, req_type);
+		limit = tot - max_qds[req_type];
+	}
+	spi->spi_req_array[req_type].sri_req_limit = limit;
+	spi->spi_req_array[req_type].sri_req_kicked = 0;
+}
+
+static int
+process_pool_cb(d_list_t *rlink, void *arg)
+{
+	struct dss_xstream	*dx = (struct dss_xstream *)arg;
+	struct sched_pool_info	*spi;
+	unsigned int		 io_max, gc_max, mig_max;
+	unsigned int		 gc_thr, mig_thr;
+	int			 rc;
+
+	spi = sched_rlink2spi(rlink);
+
+	gc_thr	= req_throttle[SCHED_REQ_GC];
+	mig_thr	= req_throttle[SCHED_REQ_MIGRATE];
+	D_ASSERT(gc_thr < 100 && mig_thr < 100);
+
+	io_max	= pool2req_cnt(spi, SCHED_REQ_IO);
+	gc_max	= pool2req_cnt(spi, SCHED_REQ_GC);
+	mig_max	= pool2req_cnt(spi, SCHED_REQ_MIGRATE);
+
+	rc = check_space_pressure(dx, spi, true);
+
+	switch (rc) {
+	case 2:	/* under severe space pressure */
+		if ((gc_max / 2) > 0)
+			io_max = min(io_max, gc_max / 2);
+		else if (gc_max)
+			io_max = min(io_max, gc_max);
+		break;
+	case 1: /* under light space pressure */
+		if (gc_max)
+			io_max = min(gc_max, io_max);
+		break;
+	case 0:	/* no space pressure, throttle gc requests */
+		if (io_max && gc_max && gc_thr) {
+			gc_thr = min(1, io_max * gc_thr / 100);
+			gc_max = min(gc_max, gc_thr);
+		}
+		break;
+	default:
+		D_ASSERT(0);
+		break;
+	}
+
+	/* Throttle rebuild and reintegraion */
+	if (mig_max && io_max && mig_thr) {
+		mig_thr = min(1, io_max * mig_thr / 100);
+		mig_max = min(mig_max, mig_thr);
+	}
+
+	reset_req_limit(dx, spi, SCHED_REQ_IO, io_max);
```

Comment:
need not differentiate update or fetch request?
Seems need not limit fetch request's handling under space pressure?

Reply:
I was thinking that both fetch and update requests need be throttled when space is under pressure, so that more CPU cycles could be used for aggregation, but I agree with you that a finer policy could be made in the future to throttle fetch and update differently.